spec/models/manageiq/providers/hawkular/middleware_manager_spec.rb

```diff
@@ -227,4 +227,107 @@ def timeline_domain_expectations(status)
       end
     end
   end
+
+  describe 'connectivity' do
+    let(:ems)      { FactoryGirl.create(:ems_hawkular) }
+    let(:username) { 'abcdefghij' }
+    let(:password) { '1234567890' }
+    let(:host)     { 'awesomehawkular.org' }
+    let(:port)     { 8080 }
+
+    describe '#connect' do
+      let(:endpoint)       { FactoryGirl.create(:endpoint) }
+      let(:http_scheme)    { URI::HTTPS }
+      let(:verify_ssl)     { 1 }
+      let(:ssl_cert_store) { nil }
+      let(:connection_options) do
+        {
+          :entrypoint  => http_scheme.build(:host => host, :port => port).to_s,
+          :credentials => { :username => username, :password => password },
+          :options     => { :tenant => 'hawkular', :verify_ssl => verify_ssl, :ssl_cert_store => ssl_cert_store }
+        }
+      end
+      subject { ems.connect }
+
+      before do
+        allow(ems).to receive_messages(
+          :hostname                => host,
+          :port                    => port,
+          :default_endpoint        => endpoint,
+          :authentication_userid   => username,
+          :authentication_password => password
+        )
+      end
+
+      context 'non-ssl' do
+        let(:http_scheme) { URI::HTTP }
+
+        it { is_expected.to be_connected_to connection_options }
+      end
+
+      context 'ssl-without-validation' do
+        let(:verify_ssl) { 0 }
+        let(:endpoint)   { FactoryGirl.create(:endpoint, :security_protocol => 'ssl-without-validation') }
+
+        it { is_expected.to be_connected_to connection_options }
+      end
+
+      context 'ssl' do
+        let(:endpoint) { FactoryGirl.create(:endpoint, :security_protocol => 'ssl') }
+
+        it { is_expected.to be_connected_to connection_options }
+      end
+    end
+
+    context 'validation' do
+      shared_examples 'validate' do
+        let(:connection) { dup }
```

Comment:
Is `dup` creating a double? I think it's the first time I see something like this.

Reply:
Nope. That's the true power of rspec. 😉  The let statement is lazy evaluated. That means this line would try evaluate the `dup` only once `connection` object is required. But this never happens in our tests. We don't need the `connection` object as is, we only need it to provide us the `inventory` method. And this can be done without being required to define what `dup` really is.

Reply:
Right. So, basically what you have here are doubles. But it's very confusing :confused: because, indeed, `dup` it's a valid method and it may just run fine and return something.

Reply:
If `connection` is defined outside, it will be available here, not sure why the dup.